### DISCUSSION TITLE
FreeDiameter as 3GPP HSS with My Own Database

### DISCUSSION THREAD

**Discussion by zimage** [freeDiameter/freeDiameter#4] [Q&A]
In the past, I extended FreeRadius, using their rlm_python extension, to look up subscriber data in our centralized CRM for authorizing subscribers to a Mikrotik-based hotspot environment. I have a similar use case in a different domain. We are running our own LTE core and the HSS has an API, but I find myself wishing our HSS could simply look up the data in realtime from our customer billing/CRM system (MS SQL Server). Is there a supported way and any example code to connect FreeDiameter to an external database for IMSI/Ki/OPC data?

**Comment by lukem** [comment]
**freeDiameter** provides an extensive plugin API framework - see http://www.freediameter.net/trac/wiki/Extensions for documentation and examples. You can implement Applications, Commands, AVPs (etc) via this extension interface.

With this extension API, it would be possible to implement a **freeDiameter** extension in C / C++ (etc) that implements the HSS S6a application for your LTE core, looking up from a database. There are extensions that already provide some of the infrastructure you'd need, including various 3GPP AVPs in `dict_dcca_3gpp.so`.
